features/crc/mbed_crc.h

```diff
@@ -0,0 +1,102 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_CRC_API_H
+#define MBED_CRC_API_H
+
+/** \addtogroup crc */
+/** @{*/
+#include <stdint.h>
+
+typedef uint64_t data_size_t;
+
+/** CRC Polynomial type
+ * 
+ * Differnet polynomials supported  
+ */
+typedef enum crc_polynomial {
+    CRC_7BIT_SD,
+    CRC_8BIT_CCITT,
+    CRC_16BIT_CCITT,
+    CRC_16BIT_SD,
+    CRC_16BIT_IBM,
+    CRC_32BIT
+} crc_polynomial_t;
+
+/** CRC width
+ * 
+ * CRC value can be 8/16/32 bit 
+ */
+typedef enum crc_width {
+    CRC_INVALID = 0,
+    CRC_7 = 7,
+    CRC_8 = 8,
+    CRC_16 = 16,
+    CRC_32 = 32,
+} crc_width_t;
+
+#define GET_DATA_SIZE(x) ((x) <= 8 ? 1 : ((x) <= 16 ? 2 : 4))
+#define TOP_BIT(x)       ((x) < 8 ? (1u << 7) : (1u << (x - 1)))
+#define DATA_MASK(x)     ((1u << x) - 1)
+
+/** CRC object provides CRC generation through hardware/software
+ */
+class CRC
+{
+public:
+    /** Lifetime of CRC object
+     */
+    virtual ~CRC() {};
+
+    /** Initialize CRC module
+     *
+     *  @return  0 on success or a negative error code on failure
+     */
+    virtual int init()
+    {
+        return 0;
+    }
+
+    /** Deinitialize CRC module
+     *
+     *  @return  0 on success or a negative error code on failure
+     */
+    virtual int deinit()
+    {
+        return 0;
+    }
+
+    /** Get CRC polynomial
+     * 
+     * Get the current CRC polynomial
+     *
+     * @return  Polynomial value
+     */
+    virtual uint32_t get_polynomial(void) const = 0;
+    
+
+    /** Compute CRC for the data input
+     *
+     *  @param  buffer  Data bytes     
+     *  @param  size  Size of data
+     *  @param  crc  CRC 
```

Comment:
Thought: you should document what the initial state of crc is expected to be.

Can I crc data without storing it all in memory at once? This is a requirement for littlefs.

Reply:
Initial state of CRC is not picked from this argument, we only return computed CRC , Initial value is don;t care. 
Inital XOR values for CRC, vary based on the CRC algorithm selected, which is set as part of constructor.

Reply:
Do you think we could have a `compute_partial` or something similar to continue a CRC without initializing the value?

Reply:
Partial CRC, yes we can have support for that in future. But partial CRC computation should be separate API and we will need proper locking mechanism with that. Partial CRC method is used to compute entire CRC in parts in parallel.

Reply:
Oh no, not for parallel computation. Just for computing a CRC when the data being CRC'd doesn't fit in RAM. The data can be streamed in, but only if the CRC can be computed partially.

This can be another pr if you'd like, it would just be needed before this class can be adopted everywhere.